### PR TITLE
Some cleanup in bytecomp/bytegen.ml 

### PR DESCRIPTION
**Pull request by hhugo**
- Remove global variables used to track exception handlers and static raise handlers (Similar to https://github.com/ocaml/ocaml/pull/11615)
- In addition, remove a global variable used to track maximum stack size for the current function. It's now tracked in a stack_info record (together with handlers)
```ocaml
type stack_info = {
  try_blocks : int list;
  (* list of stack size for each nested try block *)
  sz_static_raises : (int * (int * int * int list)) list;
  (* association staticraise numbers -> (lbl,size of stack, try_blocks *)
  max_stack_used : int ref;
  (* Maximal stack size reached during the current function body *)
}
```
- `compunit_name` was not reset when using `compile_phrase` (and could use a value set by a previous call to `compile_implementation`)
- Not longer expose a `reset` function and make sure the few remaining globals are reset before `compile_phrase`/`compile_implementation` return 

We no longer maintain the hidden invariant that upon exit from bytegen, `try_blocks` and `sz_static_raises` are reset to their initial values.

### CODE REVIEW

**Comment by hhugo** [comment]
@smuenzel, I've included and extended your commit.

**Comment by hhugo** [comment]
I've updated the PR description with some details about what it contrains

**Comment by hhugo** [comment]
@Ekdohibs, would you be interested by reviewing this PR ?

File: bytecomp/bytegen.ml [hunk 1]
```diff
@@ -352,26 +367,23 @@ let code_as_jump l sz = match l with
...
-    free_vars: Ident.t list;            (* free variables of the function *)
+  { params: Ident.t list;     (* function parameters *)
+    body: lambda;             (* the function body *)
+    label: label;             (* the label of the function entry *)
+    free_vars: Ident.t list;  (* free variables of the function *)
     num_defs: int;            (* number of mutually recursive definitions *)
-    rec_vars: Ident.t list;             (* mutually recursive fn names *)
-    rec_pos: int }                      (* rank in recursive definition *)
+    rec_vars: Ident.t list;   (* mutually recursive fn names *)
+    rec_pos: int }            (* rank in recursive definition *)
```
Comment (nojb) [inline bytecomp/bytegen.ml]: Unnecessary whitespace diff.

File: bytecomp/bytegen.ml [hunk 2]
```diff
@@ -1121,29 +1148,25 @@ let comp_remainder cont =
...
-let compile_implementation modulename expr =
-  Stack.clear functions_to_compile;
+let reset () =
   label_counter := 0;
-  sz_static_raises := [] ;
+  compunit_name := "";
+  Stack.clear functions_to_compile
+
+let compile_implementation modulename expr =
+  reset ();
```
Comment (nojb) [inline bytecomp/bytegen.ml:1157]: Is this `reset` needed given the `~finally` clause below?

File: bytecomp/bytegen.ml [hunk 3]
```diff
@@ -1121,29 +1148,25 @@ let comp_remainder cont =
...
     Kbranch lbl_init :: comp_remainder (Klabel lbl_init :: init_code)
   end else
-    init_code
+    init_code)
 
 let compile_phrase expr =
-  Stack.clear functions_to_compile;
-  label_counter := 0;
-  sz_static_raises := [] ;
+  reset ();
```
Comment (nojb) [inline bytecomp/bytegen.ml:1168]: Same question.

File: bytecomp/bytegen.ml [hunk 4]
```diff
@@ -322,28 +322,43 @@ let add_pseudo_event loc modname c =
 
 (**** Compilation of a lambda expression ****)
 
-let try_blocks = ref []  (* list of stack size for each nested try block *)
+type stack_info = {
+  try_blocks : int list;
+  (* list of stack size for each nested try block *)
+  sz_static_raises : (int * (int * int * int list)) list;
+  (* association staticraise numbers -> (lbl,size of stack, try_blocks *)
+  max_stack_used : int ref;
```
Comment (nojb) [inline bytecomp/bytegen.ml:330]: I always find the combination of immutable and mutable data in the same data structure confusing, but I believe the current use is correct, because all immutable copies of the data structure share the same `ref`. Getting rid of the ref would require `check_stack` to return a new `stack_info` and this would need propagating the new `stack_info` everywhere, which would make the change much more invasive. So I am inclined to accept the `ref` in the current version :)

**Review by nojb — APPROVED** [review]
LGTM. The new `stack_info` argument needs to be passed everywhere and is only updated in two or three places, but it gets rid of some global state which is good in the long run.

There is a small whitespace diff which would be nice to revert.

File: bytecomp/bytegen.ml (hunk 1 above)
Comment (hhugo) [reply bytecomp/bytegen.ml]: Reverted.

File: bytecomp/bytegen.ml (hunk 4 above)
Comment (hhugo) [reply bytecomp/bytegen.ml:330]: > and this would need propagating the new stack_info everywhere

In particular, the stack_info would have to be propagated up, where it's currently only propagated down.

File: bytecomp/bytegen.ml (hunk 2 above)
Comment (hhugo) [reply bytecomp/bytegen.ml:1157]: Not strictly necessary, but it seems to me it helps understanding the logic a bit.
To me, `reset` in `~finally` is about not keeping values alive for no reason. while the `reset` at the beginning is about initializing values correctly before we start using them.

**Comment by nojb** [comment]
Merged, thanks!
